applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp

```diff
@@ -0,0 +1,376 @@
+        return 0.0;
+    }
+}
+
+double LinearStrainEnergyResponseUtils::CalculateValue(const std::vector<ModelPart*>& rModelParts)
+{
+    double value = 0.0;
+    for (auto p_model_part : rModelParts) {
+        value += CalculateModelPartValue(*p_model_part);
+    }
+    return value;
+}
+
+double LinearStrainEnergyResponseUtils::CalculateModelPartValue(ModelPart& rModelPart)
+{
+    KRATOS_TRY
+
+    using tls_type = std::tuple<Matrix, Vector, Vector>;
+
+    const double local_element_value = block_for_each<SumReduction<double>>(rModelPart.Elements(), tls_type(), [&](auto& rElement, tls_type& rTLS) {
+        Matrix& r_lhs = std::get<0>(rTLS);
+        Vector& r_rhs = std::get<1>(rTLS);
+        Vector& r_u = std::get<2>(rTLS);
+        return CalculateEntityStrainEnergy(rElement, r_lhs, r_rhs, r_u, rModelPart.GetProcessInfo());
+    });
+
+    const double local_condition_value = block_for_each<SumReduction<double>>(rModelPart.Conditions(), tls_type(), [&](auto& rCondition, tls_type& rTLS) {
+        Matrix& r_lhs = std::get<0>(rTLS);
+        Vector& r_rhs = std::get<1>(rTLS);
+        Vector& r_u = std::get<2>(rTLS);
+        return CalculateEntityStrainEnergy(rCondition, r_lhs, r_rhs, r_u, rModelPart.GetProcessInfo());
+    });
+
+    return rModelPart.GetCommunicator().GetDataCommunicator().SumAll(local_element_value + local_condition_value);
+
+    KRATOS_CATCH("");
+}
+
+void LinearStrainEnergyResponseUtils::CalculateSensitivity(
+    const std::vector<ModelPart*>& rEvaluatedModelParts,
+    const SensitivityModelPartVariablesListMap& rSensitivityModelPartVariableInfo,
+    const double PerturbationSize)
+{
+    KRATOS_TRY
+
+    ResponseUtils::CheckAndPrepareModelPartsForSensitivityComputation(rEvaluatedModelParts, rSensitivityModelPartVariableInfo, SELECTED, {&SHAPE_SENSITIVITY});
+
+    // calculate sensitivities for each and every model part w.r.t. their sensitivity variables list
+    for (const auto& it : rSensitivityModelPartVariableInfo) {
+        auto& r_sensitivity_model_part = *(it.first);
+        for (auto& r_variable : it.second) {
+            std::visit([&](auto&& r_variable) {
+                if (*r_variable == YOUNG_MODULUS_SENSITIVITY) {
+                    CalculateStrainEnergyLinearlyDependentPropertySensitivity(r_sensitivity_model_part, YOUNG_MODULUS, YOUNG_MODULUS_SENSITIVITY);
+                } else if (*r_variable == THICKNESS_SENSITIVITY) {
+                    CalculateStrainEnergyLinearlyDependentPropertySensitivity(r_sensitivity_model_part, THICKNESS, THICKNESS_SENSITIVITY);
+                } else if (*r_variable == POISSON_RATIO_SENSITIVITY) {
+                    CalculateStrainEnergyFiniteDifferencePropertySensitivity(r_sensitivity_model_part, PerturbationSize, POISSON_RATIO, POISSON_RATIO_SENSITIVITY);
+                } else if (*r_variable == SHAPE_SENSITIVITY) {
+                    CalculateStrainEnergyShapeSensitivity(r_sensitivity_model_part, PerturbationSize, SHAPE_SENSITIVITY);
```

Comment:
I changed it to have the semi analytics, since it is not only FD, it is semi analytics. Would you be ok with `CalculateStrainEnergySemiAnalyticShapeSensitivity`?